torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -306,36 +316,53 @@ struct SymbolicShapeOpAnalyzer {
-      // replacing the partial evaluation of cat([x, y]) def cat(tensors:
-      // List[List[int]], dim: int)
-      //    body
-      // with
-      // def cat(x, y, dim: int)
-      //     tensors = [x, y]
-      //     body
-      uint64_t li_length = inputs_.size() - (schema_->arguments().size() - 1);
-      std::vector<Value*> li_inputs;
-      Value* graph_input = shape_compute_graph_->inputs().at(0);
-      for (size_t j = 0; j < li_length; ++j) {
-        auto new_inp = shape_compute_graph_->insertInput(j);
-        new_inp->setType(ListType::ofInts());
-        li_inputs.push_back(new_inp);
-      }
-      WithInsertPoint guard(*shape_compute_graph_->block()->nodes().begin());
-      auto new_li = shape_compute_graph_->insertNode(
-          shape_compute_graph_->createList(ListType::ofInts(), li_inputs));
-      graph_input->replaceAllUsesWith(new_li->output());
+    if (shape_compute_graph_->inputs().size() == 0) {
+      return;
+    }
 
-      shape_compute_graph_->eraseInput(li_length);
+    bool seen_tensor_list = false;
+
+    for (size_t op_in_index = 0;
+         op_in_index < shape_compute_graph_->inputs().size();
+         op_in_index++) {
+      Value* graph_in_var = shape_compute_graph_->inputs().at(op_in_index);
+      if (isListOfListOfInts(graph_in_var->type())) {
+        // Modifying the graph where _node is part of to not use the tensor
+        // construct
+
+        // When we have partially evaluate a list of Tensors like cat(tensor[])
+        // We have a few problems:
+        // - optimizing out calls to the length of the list: len(tensors)
+        // - resolving accesses of the list to the tensor symbolic sizes the
+        // corresponding list element We can solve both of these problems by
+        // replacing the partial evaluation of cat([x, y]) def cat(tensors:
+        // List[List[int]], dim: int)
+        //    body
+        // with
+        // def cat(x, y, dim: int)
+        //     tensors = [x, y]
+        //     body
+        TORCH_INTERNAL_ASSERT(
+            !seen_tensor_list,
+            "SSA doesn't handle case with multiple tensor lists")
+        seen_tensor_list = true;
+
+        uint64_t li_length = inputs_.size() - (schema_->arguments().size() - 1);
+        std::vector<Value*> li_inputs;
+
+        TypePtr element_type =
+            graph_in_var->type()->cast<ListType>()->getElementType();
+        for (size_t j = op_in_index; j < li_length; ++j) {
+          auto new_inp = shape_compute_graph_->insertInput(j);
+          new_inp->setType(element_type);
+          li_inputs.push_back(new_inp);
+        }
+        WithInsertPoint guard(*shape_compute_graph_->block()->nodes().begin());
+        auto new_li = shape_compute_graph_->insertNode(
+            shape_compute_graph_->createList(element_type, li_inputs));
+        graph_in_var->replaceAllUsesWith(new_li->output());
+        shape_compute_graph_->eraseInput(li_length);
```

Comment:
this should be `li_length + op_in_index` I think

Also I think this will skip an input if `li_length == 0`, but I guess that's not really an issue unless you have an op with multiple List[Tensor] inputs **and** the first List[Tensor] input has length 0.

Reply:
I explicitly assert that there is only one List[Tensor] above, as this all breaks down if there are two List[Tensor] objects.

Reply:
yes, but if you had two List[Tensor] arguments and the passed an input of length 0 into the first argument, you wouldn't hit the assertion

Reply:
The `if` condition checks if the shape function has the type `List[List[int]]`. The fact that an input is length 0 should not affect this.

Reply:
If we erase this input without inserting any new inputs (the case where the list is empty), then I think it will skip the second List[List[int]] input, e.g.

```
op(x: List[List[int]], y: List[List[int]], z: List[int])
   ^ op_in_index = 0, suppose input is List[Tensor] of length 0
     We don't insert anything since the length was 0
     Get to the end of the loop, and erase this input and increment op_in_index

op([removed], y: List[List[int]], z: List[int])
                                  ^ op_in_index = 1, we skipped the second List[List[int]]
                                    in the second iteration through the loop
```